Reimplement per-operator selective build

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39401 Reimplement per-operator selective build**

This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

We don't delete the old selective build as it is currently still
used internally for other reasons that aren't subsumed here.
However, this mechanism can now be used to handle non code generated
registrations IF they are placed within the registration library
that is recompiled per mobile application.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

ezyang committed:
Reimplement per-operator selective build

This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR does not have feature parity with the old selective build
apparatus.  These things are known to be missing:

- Prevent registration of SparseCPU and MkldnnCPU implementations
  (backend_whitelist)
- Prevent schema def() entirely in some situations
  (force_schema_registration)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/op_registration/op_whitelist.h:
```diff
@@ -0,0 +1,31 @@
+ * set of operators (+ dependencies) included.
+ * - Build with -DTORCH_OPERATOR_WHITELIST="aten::add;aten::sub" and only these
+ *   two ops will be included in your build.  NB: the whitelist does NOT contain
+ *   overload names.
```
yes

bhosmer replied:
Might be worth adding to the comment? Don't know if I'm an outlier here in terms of liking that kind of stuff totally spelled out, your call (also LMK if this is jumping the gun on review-type suggestions)

bhosmer commented on c10/util/constexpr_string_functions.h:
```diff
@@ -0,0 +1,81 @@
+// as looping constructs (like for) are not constexpr.  This code is
+// written using ternaries as a holdover from C++11; but as of C++14
+// constexpr functions are allowed to have multiple return statements
+// (writing them with ternaries is more portable though!)
```
Quick thought, is there an existing API for this kind of stuff we could emulate, like in Boost or something? (maybe you are already?)

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR does not have feature parity with the old selective build
apparatus.  These things are known to be missing:

- Prevent registration of SparseCPU and MkldnnCPU implementations
  (backend_whitelist)
- Prevent schema def() entirely in some situations
  (force_schema_registration)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

smessmer commented on torch/library.h:
```diff
@@ -202,6 +203,46 @@ namespace detail {
+    constexpr SelectiveName(const char* name) : name_(name) {}
+    constexpr operator const char*() { return name_; }
+  private:
+    const char* name_;
```
We should use c10::string_view here and also rewrite the constexpr string functions to use it

ezyang replied:
Mm, string view here isn't really necessary. Is there any particular reason you want a string view here?

smessmer replied:
just generally prefer string_view to const char* because it is more flexible and doesn't have downsides. Say at some point you want to instantiate a SelectiveStr holding only the op name from a `const char*` holding the full schema, `string_view` would easily allow that while `const char*` has to be null terminated. But not really important here.

smessmer reviewed (dismissed):
generally looks very good. Requesting changes because I want to take another look once it's using c10::string_view

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR does not have feature parity with the old selective build
apparatus.  These things are known to be missing:

- Prevent registration of SparseCPU and MkldnnCPU implementations
  (backend_whitelist)
- Prevent schema def() entirely in some situations
  (force_schema_registration)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
I updated the code to use `string_view`, so the whitelist constexpr stuff is shorter (though it smacks of the kind of C-era string manipulation that would be better not to do)

bhosmer commented on c10/util/constexpr_string_functions.h:
```diff
@@ -0,0 +1,32 @@
+
+// returns true iff whitelist contains item
+// op_whitelist_contains("a;bc;d", "bc") == true
+constexpr bool op_whitelist_contains(string_view whitelist, string_view item) {
```
This is a pretty op-whitelist-specific piece of functionality to be in `c10/util/constexpr_string_functions.h`, no? Why not just put it in `op_whitelist.h`?

OTOH if you're thinking it's the first of a more general-purpose crop of utilities, I think we'd want to rename it to like `semi_delimited_list_contains` (and arguably just put it in `string_view.h`, seems like there's a good amount of this kind of stuff in there already)

ezyang replied:
There's no good reason; @smessmer's original patch had this in its own header and I've been to lazy to move it.

bhosmer reviewed (approved):
+1, really nice and clear and well explicated. One minor suggestion inline, and also echoing Sebastian's comment re string_view

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
There are some dangerous aspects to this PR.

* Suppose I write `m.impl(TORCH_SELECTIVE_NAME(add), &add)` instead of `TORCH_SELECTIVE_NAME(aten::add)`. Then I will always disable this operator on mobile builds. I think adding a `assert(item.find("::") != string_view::npos);` should help; I'm currently testing to make sure this is the case and will add this to the PR if so.
* The need to replicate namespace for selective build is a bit troublesome. It would be nice to not have to replicate it. However, it seems like it is a bit difficult to do this under the current design. `TORCH_SELECTIVE_NAME` operates without any reference to m, so if you omit the namespace there is no reasonable way to determine what the namespace should be. One way around this is I could define a local constexpr variable like `__the_namespace` and then `TORCH_SELECTIVE_NAME` can tack it onto the operator when it exists, except with the current `TORCH_LIBRARY` macro design I don't have anywhere I can actually define this variable.
* In some situations in the current API, we aren't able to statically tell what the dispatch key / operator name is. For example, if I say `m.impl("aten::add", torch::dispatch(kCPU, add_cpu))`, the CPU key here isn't statically known. This PR will default to including operators in that case. One way to make this a little safer is to add a runtime test and exclude the operator in that case (or raise a warning saying that N is disabled, but we weren't able to prune it from the build). I haven't implemented this yet, looking for some opinions.
* I am wondering if we should reduce the API surface of torch/library.h so that people are only able to specify usage patterns for which we can statically work things out. So for example, remove `torch::dispatch` from the public API surface (and force people to always use `TORCH_LIBRARY_IMPL` instead). Similarly, given the problems with contextual namespaces, maybe we should just force namespaces to be redundantly specified. :/

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ljk53 commented on tools/code_analyzer/op_deps_pass.cpp:
```diff
@@ -744,8 +744,12 @@ class OpDependency : public ModulePass {
+      // a namespace is included: a double colon will be present.
       const std::string& schemaStr =
-          contextualNamespace.empty() ? str : contextualNamespace + str;
+          (contextualNamespace.empty() || str.find("::") != std::string::npos) ? str : contextualNamespace + str;
```
lgtm - just to confirm, so contextual namespace and operator namespace will both exist after moving to the new world?

ezyang replied:
Yes, although it would be nice if the constexpr machinery could get the contextual namespace so I could get rid of this https://github.com/pytorch/pytorch/pull/39401#issuecomment-639599944

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

With this new implementation, we remove code generated whitelisting,
generate to separate files

- Revert "[pytorch] move force schema registration output into a separate file (#36284)"
  This reverts commit 126d00c8ddcca996a5045f6a6c59cf86c120936c.
- Revert "[pytorch] force c10 schema registration for custom build"
  This reverts commit 61b680c01250c420603ae2937a2f2f16adebc475.
- Revert "[pytorch] aten codegen to filter backends for default mobile build"
  This reverts commit 6792bafa7258e66cf66f738fac6ab8316c4bc0f8
- Revert "[pytorch] codegen flags to whitelist op registrations / generate to separate files (#32451)"
  This reverts commit 69283388caa1fabe695ca0209ea03bed830de9be.

The new plan on record is that we will control whether or not
we should register an operator or not by a non-codegenned mechanism
(constexpr conditional).

This PR doesn't actually implement this new mechanism yet (for ease
of review), which means it will break all of our selective builds.
I intend to squash it with a future PR later.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

With this new implementation, we remove code generated whitelisting,
generate to separate files

- Revert "[pytorch] move force schema registration output into a separate file (#36284)"
  This reverts commit 126d00c8ddcca996a5045f6a6c59cf86c120936c.
- Revert "[pytorch] force c10 schema registration for custom build"
  This reverts commit 61b680c01250c420603ae2937a2f2f16adebc475.
- Revert "[pytorch] aten codegen to filter backends for default mobile build"
  This reverts commit 6792bafa7258e66cf66f738fac6ab8316c4bc0f8
- Revert "[pytorch] codegen flags to whitelist op registrations / generate to separate files (#32451)"
  This reverts commit 69283388caa1fabe695ca0209ea03bed830de9be.

The new plan on record is that we will control whether or not
we should register an operator or not by a non-codegenned mechanism
(constexpr conditional).

This PR doesn't actually implement this new mechanism yet (for ease
of review), which means it will break all of our selective builds.
I intend to squash it with a future PR later.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

With this new implementation, we remove code generated whitelisting,
generate to separate files

- Revert "[pytorch] move force schema registration output into a separate file (#36284)"
  This reverts commit 126d00c8ddcca996a5045f6a6c59cf86c120936c.
- Revert "[pytorch] force c10 schema registration for custom build"
  This reverts commit 61b680c01250c420603ae2937a2f2f16adebc475.
- Revert "[pytorch] aten codegen to filter backends for default mobile build"
  This reverts commit 6792bafa7258e66cf66f738fac6ab8316c4bc0f8
- Revert "[pytorch] codegen flags to whitelist op registrations / generate to separate files (#32451)"
  This reverts commit 69283388caa1fabe695ca0209ea03bed830de9be.

The new plan on record is that we will control whether or not
we should register an operator or not by a non-codegenned mechanism
(constexpr conditional).

This PR doesn't actually implement this new mechanism yet (for ease
of review), which means it will break all of our selective builds.
I intend to squash it with a future PR later.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang commented:
@ljk53 do you know how I would go about debugging the ios failure? It looks legit but doesn't have any information

ljk53 commented:
> @ljk53 do you know how I would go about debugging the ios failure? It looks legit but doesn't have any information

I'd try to run it in xcode. @xta0 do you have some instruction to repro ios failure locally? and/or can we improve the error message to surface more details?

ezyang commented:
BTW, in fbcode, `//Apps/Internal/PyTorchPlayground:PyTorchPlaygroundTests - PyTorchFBNetTests/testFBNetModel (macos, iphonesimulator)` failed the same way, so I'm betting it's probably the same way (also no real error message)

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

With this new implementation, we remove code generated whitelisting,
generate to separate files

- Revert "[pytorch] move force schema registration output into a separate file (#36284)"
  This reverts commit 126d00c8ddcca996a5045f6a6c59cf86c120936c.
- Revert "[pytorch] force c10 schema registration for custom build"
  This reverts commit 61b680c01250c420603ae2937a2f2f16adebc475.
- Revert "[pytorch] aten codegen to filter backends for default mobile build"
  This reverts commit 6792bafa7258e66cf66f738fac6ab8316c4bc0f8
- Revert "[pytorch] codegen flags to whitelist op registrations / generate to separate files (#32451)"
  This reverts commit 69283388caa1fabe695ca0209ea03bed830de9be.

The new plan on record is that we will control whether or not
we should register an operator or not by a non-codegenned mechanism
(constexpr conditional).

This PR doesn't actually implement this new mechanism yet (for ease
of review), which means it will break all of our selective builds.
I intend to squash it with a future PR later.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

We don't delete the old selective build as it is currently still
used internally for other reasons that aren't subsumed here.
However, this mechanism can now be used to handle non code generated
registrations IF they are placed within the registration library
that is recompiled per mobile application.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

We don't delete the old selective build as it is currently still
used internally for other reasons that aren't subsumed here.
However, this mechanism can now be used to handle non code generated
registrations IF they are placed within the registration library
that is recompiled per mobile application.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

ezyang commented:
I pushed an update to just have the implementation bits, but not cut anything over; this unblocks @iseeyuan and also is immediately landable, since previously this was blocked on fbcode internal mobile problems.

ezyang committed:
Update on "Reimplement per-operator selective build"


This uses the technique proposed by smessmer in D16451848 to selectively
register operators without codegen.  See the Note inside for more
details.

This PR has feature parity with the old selective build apparatus:
it can whitelist schema def()s, impl()s, and on a per dispatch key
basis.  It has expanded dispatch key whitelisting, whereas previously
manually written registrations were not whitelisted at all.  (This
means we may be dropping dispatch keys where we weren't previously!)

We don't delete the old selective build as it is currently still
used internally for other reasons that aren't subsumed here.
However, this mechanism can now be used to handle non code generated
registrations IF they are placed within the registration library
that is recompiled per mobile application.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21905593](https://our.internmc.facebook.com/intern/diff/D21905593)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/operator_name.h:
```diff
@@ -45,6 +45,25 @@ struct OperatorName final {
+// Non-owning view of an OperatorName.  Unlike OperatorName, most of
+// its functions are constexpr, so it can be used for compile time
+// computations
+struct OperatorNameView final {
```
test cases

smessmer commented on torch/library.h:
```diff
@@ -298,6 +299,48 @@ namespace detail {
+  // directly; use TORCH_SELECTIVE_NAME or TORCH_SELECTIVE_SCHEMA macros below
+  // to create it.
+  template <bool enabled>
+  class SelectiveStr {
```
Btw, this is an amazing idea and I somehow had forgotten about it. I was trying to find something that avoids having to type the schema string twice at the call site a couple days ago when talking with Martin about selective build for prim ops but didn't come up with this solution :)

smessmer commented on torch/library.h:
````diff
@@ -435,8 +478,8 @@ class CAFFE2_API Library final {
   /// ```
-  template <typename Func>
-  Library& impl(const char* name, Func&& raw_f) & {
+  template <typename Name, typename Func>
````
This pattern opens up `impl` to be called with any type for `Name` and will yield very bad error messages when called with non-string types or string types that we don't support. Can we keep the old implementation (or maybe actually change it to take string_view instead of const char*) and then explicitly specialize for `SelectiveStr<true>`? The specialization could just forward to the `string_view` version.
